File: sample/iosSwiftApp/iosSwiftApp/PickersScreen.swift
//
//  PickersScreen.swift
//  Nepali Date Picker
//
//  The full calendar picker, including a pre-selected date and the paired Gregorian variant.
//

import SwiftUI
import nepali_date_picker

struct PickersScreen: View {
    @State private var plain: CustomCalendar?
    @State private var preselected: CustomCalendar?
    @State private var withEnglish: CustomCalendar?
    @State private var nepaliLocale: CustomCalendar?
    @State private var range: (CustomCalendar?, CustomCalendar?) = (nil, nil)

    /// A fixed date so the pre-selection demo does not move with the calendar.
    private let preselectedDate = SimpleDate(year: 2081, month: 1, dayOfMonth: 15)

    var body: some View {
        ScrollView {
            VStack(spacing: 16) {
                DemoSection(
                    title: "Default picker",
                    subtitle: "No initial selection. The mode toggle switches between calendar and typed input."
                ) {
                    AutoSized(measurementHeight: 560) { report in
                        NepaliDatePickerView(onHeightChange: report, initialSelectedDate: nil) { plain = $0 }
                    }
                    SelectionSummary(selection: plain)
                }

                DemoSection(
                    title: "Pre-selected date",
                    subtitle: "Opens on Baisakh 15, 2081 with that date already chosen and its month displayed."
                ) {
                    AutoSized(measurementHeight: 560) { report in
                        NepaliDatePickerView(onHeightChange: report, initialSelectedDate: preselectedDate) { preselected = $0 }
                    }
                    SelectionSummary(selection: preselected)
                }

                DemoSection(
                    title: "With the Gregorian date",
                    subtitle: "Every Bikram Sambat day is annotated with its English equivalent."
                ) {
                    AutoSized(measurementHeight: 620) { report in
                        NepaliDatePickerView(
                            onHeightChange: report,
                            initialSelectedDate: preselectedDate,
                            showEnglishDate: true
                        ) { withEnglish = $0 }
                    }
                    SelectionSummary(selection: withEnglish)
                    if let english = withEnglish {
                        LabeledValue(
                            label: "Gregorian",
                            value: NepaliDateConverter.shared.convertNepaliToEnglish(
                                nepaliYYYY: english.year,
                                nepaliMM: english.month,
                                nepaliDD: english.dayOfMonth
                            ).text
                        )
                    }
                }

                DemoSection(
                    title: "Nepali language",
                    subtitle: "The same picker localized, with Devanagari month names and digits."
                ) {
                    AutoSized(measurementHeight: 560) { report in
                        NepaliDatePickerView(
                            onHeightChange: report,
                            initialSelectedDate: nil,
                            locale: SampleDefaults.nepali
                        ) { nepaliLocale = $0 }
                    }
                    SelectionSummary(selection: nepaliLocale)
                }

                DemoSection(
                    title: "Range picker",
                    subtitle: "Pick a start and an end. Months are stacked vertically."
                ) {
                    AutoSized(measurementHeight: 640) { report in
                        NepaliDateRangePickerView(
                            onHeightChange: report,
                            initialSelectedStartDate: preselectedDate,
                            initialSelectedEndDate: SimpleDate(year: 2081, month: 1, dayOfMonth: 25),
                            locale: SampleDefaults.englishRange
                        ) { start, end in range = (start, end) }
                    }
                    LabeledValue(label: "Start", value: range.0?.text ?? "none")
                    LabeledValue(label: "End", value: range.1?.text ?? "none")
                }

                DemoSection(
                    title: "Range with Gregorian dates",
                    subtitle: "The range calendar paired with English dates and horizontal paging."
                ) {
                    AutoSized(measurementHeight: 640) { report in
                        NepaliDateRangePickerView(
                            onHeightChange: report,
                            initialSelectedStartDate: nil,
                            initialSelectedEndDate: nil,
                            locale: SampleDefaults.englishRange,
                            showMonthsVertically: false,
                            showEnglishDate: true
                        ) { _, _ in }
                    }
                }
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 16)
        }
        .navigationTitle("Pickers")
    }
}

File: sample/flutterApp/lib/src/screens/pickers_screen.dart
// Copyright © 2026 Shiva Thapa (@shivathapaa). All rights reserved.
// Use of this source code is governed by the MPL 2.0 license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:nepali_date_picker_kmp/nepali_date_picker_kmp.dart';

import '../support/demo_section.dart';
import '../support/sample_defaults.dart';

/// Single, localized, dual-date and range grids, matching the sibling
/// showcases' Pickers screen section for section.
class PickersScreen extends StatefulWidget {
  const PickersScreen({super.key});

  @override
  State<PickersScreen> createState() => _PickersScreenState();
}

class _PickersScreenState extends State<PickersScreen> {
  String _defaultPick = 'none';
  String _preselectedPick = 'none';
  String _dualPick = 'none';
  String _dualGregorian = 'none';
  String _nepaliPick = 'none';
  String _rangeStart = 'none';
  String _rangeEnd = 'none';
  String _dualRangeStart = 'none';
  String _dualRangeEnd = 'none';

  Future<void> _showGregorian(NepaliDate? date) async {
    if (date == null) return;
    final gregorian = await NepaliDateConverter.convertBsToAd(
        date.year, date.month, date.dayOfMonth);
    if (!mounted) return;
    setState(() => _dualGregorian = calendarText(gregorian));
  }

  @override
  Widget build(BuildContext context) => DemoScreen(
        title: 'Pickers',
        children: [
          DemoSection(
            title: 'Default picker',
            subtitle: 'No initial selection. The mode toggle switches between '
                'calendar and typed input.',
            child: Column(children: [
              NepaliDatePicker(
                onDateSelected: (date) =>
                    setState(() => _defaultPick = calendarText(date)),
              ),
              SelectionSummary(_defaultPick),
            ]),
          ),
          DemoSection(
            title: 'Pre-selected date',
            subtitle: 'Opens on Baisakh 15, 2081 with that date already '
                'chosen and its month displayed.',
            child: Column(children: [
              NepaliDatePicker(
                initialSelectedDate: SampleDefaults.preselectedDate,
                onDateSelected: (date) =>
                    setState(() => _preselectedPick = calendarText(date)),
              ),
              SelectionSummary(_preselectedPick),
            ]),
          ),
          DemoSection(
            title: 'With the Gregorian date',
            subtitle: 'Every day carries its Gregorian equivalent, and the '
                'headline shows both calendars.',
            child: Column(children: [
              NepaliDatePicker(
                initialSelectedDate: SampleDefaults.preselectedDate,
                showEnglishDate: true,
                onDateSelected: (date) {
                  setState(() => _dualPick = calendarText(date));
                  _showGregorian(date);
                },
              ),
              SelectionSummary(_dualPick),
              LabeledValue('Gregorian', _dualGregorian),
            ]),
          ),
          DemoSection(
            title: 'Nepali language',
            subtitle: 'The same picker, localized: Nepali month and weekday '
                'names with Devanagari digits.',
            child: Column(children: [
              NepaliDatePicker(
                locale: SampleDefaults.nepali,
                onDateSelected: (date) =>
                    setState(() => _nepaliPick = calendarText(date)),
              ),
              SelectionSummary(_nepaliPick),
            ]),
          ),
          DemoSection(
            title: 'Range picker',
            subtitle: 'A ten-day range pre-selected, months stacked '
                'vertically the Material way.',
            child: Column(children: [
              NepaliDateRangePicker(
                initialSelectedStartDate: SampleDefaults.preselectedDate,
                initialSelectedEndDate: const SimpleDate(2081, 1, 25),
                locale: SampleDefaults.englishRange,
                onRangeSelected: (start, end) => setState(() {
                  _rangeStart = calendarText(start);
                  _rangeEnd = calendarText(end);
                }),
              ),
              LabeledValue('Start', _rangeStart),
              LabeledValue('End', _rangeEnd),
            ]),
          ),
          DemoSection(
            title: 'Range with Gregorian dates',
            subtitle: 'Horizontal paging, and both calendars on every day.',
            child: Column(children: [
              NepaliDateRangePicker(
                showMonthsVertically: false,
                showEnglishDate: true,
                onRangeSelected: (start, end) => setState(() {
                  _dualRangeStart = calendarText(start);
                  _dualRangeEnd = calendarText(end);
                }),
              ),
              LabeledValue('Start', _dualRangeStart),
              LabeledValue('End', _dualRangeEnd),
            ]),
          ),
        ],
      );
}
